### PR TITLE
Add routing on message properties

### PR DESCRIPTION
**Pull request by gmr**
Summary:

Add a new ``hash-property`` argument setting that allows for message hasing based upon the ``correlation_id`` or ``message_id``.

Changes:

- Validate the exchange upon creation to ensure that ``hash-header`` and ``hash-property`` are not both set at the same time. Additionally validate the value of ``hash-property`` when set is one of ``correlation_id`` or ``message_id``
- Validate the routing key when adding a binding to ensure it is an integer value. In addition, validate that the value passed in is > 0 addressing #6
- Change the signature of ``hash/2`` for header to match on ``{header, Header}`` instead of ``{longstr, Header}``
- Add a new ``hash/2`` implementation that matches on ``{property, Value}`` for returning the hashable string from the message properties
- Implement a new ``hash_on/1`` method for selecting the data the message will be routed on
- Implement a new ``hash_args/1`` method for returning the configuration for both ``hash-header`` and ``hash-property``
- Add test coverage for message property based routing

This addresses the proposal I outlined in #7

### CODE REVIEW

**Comment by gmr** [comment]
FYI: I have sent in an updated contributor agreement.

**Comment by michaelklishin** [comment]
I confirm that we have a RabbitMQ CLA from @gmr on file now.

File: src/rabbit_exchange_type_consistent_hash.erl [hunk 1]
```diff
@@ -84,9 +85,34 @@ route(#exchange { name      = Name,
...
+        true  -> ok;
+        false ->
+          rabbit_misc:protocol_error(precondition_failed,
+                                     "Unsupported property: ~s",
+                                     [Value])
+      end;
+    {_, undefined} -> ok;
+    {_, _} ->
+      rabbit_misc:protocol_error(precondition_failed,
+                                 "hash-header and hash-property are mutually exclusive",
```
Comment (videlalvaro) [inline src/rabbit_exchange_type_consistent_hash.erl]: do you think you could add a test for the failure cases like here: https://github.com/rabbitmq/rabbitmq-delayed-message-exchange/blob/master/test/src/rabbit_exchange_type_delayed_message_test.erl#L33 ?

File: src/rabbit_exchange_type_consistent_hash.erl [hunk 2]
```diff
@@ -159,9 +185,34 @@ find_numbers(Source, N, Acc) ->
...
+hash({property, Property}, #basic_message {content = Content }) ->
+    #content{properties = #'P_basic'{correlation_id = CorrId, message_id = MsgId }} =
+        rabbit_binary_parser:ensure_content_decoded(Content),
+    case Property of
+      <<"correlation_id">> -> CorrId;
+      <<"message_id">> -> MsgId
+    end.
+
+hash_args(Args) ->
+    Header = case rabbit_misc:table_lookup(Args, <<"hash-header">>) of
```
Comment (videlalvaro) [inline src/rabbit_exchange_type_consistent_hash.erl]: here you can move the `case` to the next line, and then indent the whole case block

**Comment by videlalvaro** [comment]
I've added some comments. For indentation we use Emacs style. The same for paragraph formatting. We use `M-x fill-region` which is the same as select paragraph and then `M-q`.

**Comment by gmr** [comment]
Not being an emacs user, are those defaults? I'm using a "reformat with emacs" plugin in my IDE and it seems like it's only changing my code, which I'd imagine means it is.

**Comment by michaelklishin** [comment]
@gmr I'm not sure I understand that. If you need help with re-formatting your changes, I'm happy to help with that.

**Comment by videlalvaro** [comment]
BTW, by Emacs formatting I meant: emacs Erlang mode formatting, but if this is too much of a hassle we can do it for you.

File: src/rabbit_exchange_type_consistent_hash.erl [hunk 3]
```diff
@@ -84,9 +85,34 @@ route(#exchange { name      = Name,
...
+            end;
+        {_, undefined} -> ok;
+        {_, _} ->
+            rabbit_misc:protocol_error(precondition_failed,
+                                       "hash-header and hash-property are mutually exclusive",
+                                       [])
+    end.
 
-validate_binding(_X, _B) -> ok.
+validate_binding(_X, #binding { key = K }) ->
```
Comment (videlalvaro) [inline src/rabbit_exchange_type_consistent_hash.erl]: If this change and related tests is related to #6 then it would be nice to have independent PRs for that. I think fixing the negative routing key issue could even go into 3.5.5. What do you think @michaelklishin ? 

Sorry I didn't spot this earlier, but we usually pack one fix per PR

Comment (gmr) [inline src/rabbit_exchange_type_consistent_hash.erl]: Sure, I'll close this out and make a new PR for the validation of the routing key. Once merged, I'll submit a PR based on that adding the ``hash-property``

Comment (gmr) [inline src/rabbit_exchange_type_consistent_hash.erl]: Oh and re-formatting, the Erlang plugin I use in IntelliJ seems to do exactly what you want with its "Reformat with Emacs" command.

**Comment by videlalvaro** [comment]
About the PR, the routing key fix should be against stable branch, so we can release the fix for 3.5.5.

The new properties hashing should go against master branch. Thanks

**Comment by gmr** [comment]
Does master get rebased off of stable or should it go to both?

**Comment by videlalvaro** [comment]
First we'll merge your PR to stable (provided it is submitted against stable), and then we take care of merging those changes into master as well.
